crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,261 @@
+        let (manifest, digest) = self.oci.pull_image_manifest(&reference, &auth).await?;
+
+        let manifest_json = serde_json::to_string(&manifest)?;
+        tracing::debug!("Pulled manifest: {}", manifest_json);
+
+        // Write the manifest in `<cache_root>/registry/oci/manifests/repository:<tag_or_latest>/manifest.json`
+        let m = self
+            .cache
+            .manifest_for_reference(&reference.to_string())
+            .await?;
+        fs::write(&m, &manifest_json).await?;
+
+        let mut cfg_bytes = Vec::new();
+        self.oci
+            .pull_blob(&reference, &manifest.config.digest, &mut cfg_bytes)
+            .await?;
+        let cfg = std::str::from_utf8(&cfg_bytes)?;
+        tracing::debug!("Pulled config: {}", cfg);
+
+        // Write the config object in `<cache_root>/registry/oci/manifests/repository:<tag_or_latest>/config.json`
+        let c = self
+            .cache
+            .config_for_reference(&reference.to_string())
+            .await?;
+        fs::write(&c, &cfg).await?;
+
+        // If a layer is a Wasm module, write it in the Wasm directory.
+        // Otherwise, write it in the data directory.
+        for layer in manifest.layers {
+            // Skip pulling if the digest already exists in the wasm or data directories.
+            if self.cache.wasm_file(&layer.digest).is_ok()
+                || self.cache.data_file(&layer.digest).is_ok()
+            {
+                tracing::debug!("Layer {} already exists in cache", &layer.digest);
+                continue;
+            }
+            tracing::debug!("Pulling layer {}", &layer.digest);
+            let mut bytes = Vec::new();
+            self.oci
+                .pull_blob(&reference, &layer.digest, &mut bytes)
+                .await?;
+
+            match layer.media_type.as_str() {
+                WASM_LAYER_MEDIA_TYPE => self.cache.write_wasm(&bytes, &layer.digest).await?,
+                _ => self.cache.write_data(&bytes, &layer.digest).await?,
+            }
+        }
+
+        tracing::info!("Pulled {}@{}", reference, digest);
+
+        Ok(())
+    }
+
+    /// Create a new wasm layer based on a file.
+    pub async fn wasm_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading wasm module from {:?}", file);
+        Ok(ImageLayer::new(
+            fs::read(file).await.context("cannot read wasm module")?,
+            WASM_LAYER_MEDIA_TYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Create a new data layer based on a file.
+    pub async fn data_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading data file from {:?}", file);
+        Ok(ImageLayer::new(
+            fs::read(&file).await?,
+            DATA_MEDIATYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Construct the registry authentication based on the reference.
+    pub fn auth(reference: &Reference) -> Result<RegistryAuth> {
```

Comment:
Do this and `build_config` need to be `pub`?  It feels like that stuff should be encapsulated in the `Client`.